doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
+
+The implementation is greatly simplified by the logical replication protocol
+that guarantees that we'll receive full transactions in commit order. In the
+replication stream each transaction is represented by a BEGIN message, followed
+by one or more row updates, and finally by a COMMIT message.
+
+Each transaction is timestamped with the wall-clock time at the time of the
+BEGIN message. This allows the source implementation to emit rows as it receives
+them and ensure that transaction boundaries are guaranteed.
+
+### Adaptation to multiple tables
+
+If the sources naively applied the mechanism described above for every source
+instance they would very quickly exhaust the available replication slots from
+upstream and also multiply the bandwidth used by several times.
+
+In order to have one independent `SOURCE` object for each upstream table while
+also maintaining a single replication connection between them a new SQL syntax
+will be introduced that allows creating multiple sources at once.
+
+Specifically, a user should be able to do `CREATE SOURCES FROM POSTGRES..` and
+have one `SOURCE` object created for each upstream table. By introducing this
+syntax the SQL planner will have an opportunity to create all the relevant
+sources at once and therefore give them some shared state. The shared state
+could be some form of `Arc<RwLock<Option<T>>>` where the first source instance
+that grabs the write lock will initialize it and afterwards all the source
+instances will keep a read lock around to share the stream.
+
+The concrete syntax for `CREATE SOURCES` can be seen in
+https://github.com/MaterializeInc/materialize/pull/6299.
+
+## Open questions
+
+### Robust resume logic
+
+A strategy for handling connection disconnects is missing for both the initial
+syncing phase and the streaming phase.
+
+The initial syncing phase is the most challenging one but also the least
+important one as this only happens on source startup. In order to handle
+disconnections during this phase the implementation needs to be able to issue a
+brand new COPY command at the same point as the interrupted transfer. Normally,
+a system could just flush its in-flight state and start anew, but since we're
+streaming data into a differential dataflow this isn't possible. In order for
+the source to start over it first needs to issue retractions for all the data,
+which is what we're trying to get to in the first place.
+
+There exist subcases of this problem that make it easier. For example, upstream
+tables with an auto-increment primary key have a natural way of resuming form
+where we left off, just remember the last id. Finally, a naive implementation
+could buffer the initial snapshot in-memory, outside of the dafalow, and only
+issue the data into the dataflow once we've received a full snapshot.
```

Comment:
How would this work exactly? Once we receive the full snapshot, we read all of the dataflow and retract each point, and then insert all the points from the snapshot?

Reply:
No, it would be the other way around. In the initial sync we buffer everything in memory in the source and don't issue anything in the dataflow.

If the connection is interrupted we can just throw away our buffered data and start from the beginning.

Once we receive a full snapshot we can issue data in the dataflow and not worry about disconnections again because even if they happen we now have the replication slot that supports resumption